Update for newer AbstractMCMC/Turing interface

penelopeysm opened the pull request:
Comments in PR!

penelopeysm committed:
Update for newer AbstractMCMC/Turing interface

penelopeysm commented on ext/SliceSamplingTuringExt.jl:
```diff
@@ -6,35 +6,6 @@ using Random
 using SliceSampling
 using Turing
 
-# Required for using the slice samplers as `externalsampler`s in Turing
-# begin
-function Turing.Inference.getparams(
-    ::Turing.DynamicPPL.Model, sample::SliceSampling.Transition
-)
-    return sample.params
-end
-# end
```
The changes to the external sampler interface are described in https://github.com/TuringLang/Turing.jl/releases/tag/v0.42.0 -- the general aim is that you should not need to overload Turing internal functions (getparams was actually not exported afaik) and shifting this to AbstractMCMC means that it's easier for other packages to make use of this info.

`Turing.Inference.getparams` is gone now, it's replaced with `AbstractMCMC.getparams` (but called on the state instead of the transition).

penelopeysm commented on ext/SliceSamplingTuringExt.jl:
```diff
@@ -6,35 +6,6 @@ using Random
-
-# Required for using the slice samplers as `Gibbs` samplers in Turing
-# begin
-Turing.Inference.isgibbscomponent(::SliceSampling.RandPermGibbs) = true
-Turing.Inference.isgibbscomponent(::SliceSampling.HitAndRun) = true
-Turing.Inference.isgibbscomponent(::SliceSampling.Slice) = true
-Turing.Inference.isgibbscomponent(::SliceSampling.SliceSteppingOut) = true
-Turing.Inference.isgibbscomponent(::SliceSampling.SliceDoublingOut) = true
```
`isgibbscomponent` now defaults to true so it no longer needs to be overridden

penelopeysm commented on src/SliceSampling.jl:
```diff
@@ -37,6 +38,22 @@ struct Transition{P,L<:Real,I<:NamedTuple}
+
+Base type for MCMC states that contain a `Transition` stored in the `transition` field.
+"""
+abstract type AbstractStateWithTransition end
+AbstractMCMC.getparams(state::AbstractStateWithTransition) = state.transition.params
+AbstractMCMC.getstats(state::AbstractStateWithTransition) = state.transition.info
+function AbstractMCMC.setparams!!(
+    model::AbstractMCMC.LogDensityModel, state::AbstractStateWithTransition, params
+)
+    new_lp = LogDensityProblems.logdensity(model.logdensity, params)
+    new_transition = Transition(params, new_lp, NamedTuple())
+    return Accessors.@set state.transition = new_transition
+end
```
Since the definitions of these functions are the same for all states in this package, I thought it would be cleaner to just define the behaviour on an abstract type. It does necessitate an extra dep on Accessors, but that's fairly lightweight.

penelopeysm commented:
The failing tests are to do with `initial_sample` here

https://github.com/TuringLang/SliceSampling.jl/blob/2dd671699f9357033928bc4864e549187a344fa1/ext/SliceSamplingTuringExt.jl#L38-L41

The fact is that when SliceSampling is used via Turing's externalsampler mechanism, that code path will never be hit, because externalsampler makes sure to provide a vector of initial params:

https://github.com/TuringLang/Turing.jl/blob/4dc7ad096f92fd571de312e7751986484ac6cb50/src/mcmc/external_sampler.jl#L151-L177

So this code path will only be hit if somebody actually manually constructs a LogDensityFunction and passes it to SliceSampling without specifying initial params. One option is therefore to delete the function and the tests. Even though this seems like a bit of a niche use case, I also don't see any justifiable reason to _not_ support it, and deleting it seems overkill.

If we don't go down that route, we need a patch to make it work again, since DynamicPPL 0.39 removed some functionality that was needed for this. That patch is here: https://github.com/TuringLang/DynamicPPL.jl/pull/1178. Thus, we need to wait for that to be merged, and I can revisit this after that.

Once that is done, I think that the Turing ext can be changed into a DynamicPPL ext, which will make life easier for everyone (more accurate compat bounds, no need to update when there are unrelated changes in Turing, etc.)

Red-Portal commented:
> So this code path will only be hit if somebody actually manually constructs a LogDensityFunction and passes it to SliceSampling without specifying initial params. One option is therefore to delete the function and the tests. Even though this seems like a bit of a niche use case, I also don't see any justifiable reason to not support it, and deleting it seems overkill.

I think it is fine to remove the custom initialization. After all, the purpose of that `ext` is to make things work automatically with Turing, nothing more. 

> Once that is done, I think that the Turing ext can be changed into a DynamicPPL ext, which will make life easier for everyone (more accurate compat bounds, no need to update when there are unrelated changes in Turing, etc.)

Yes, sounds good to me. Let me handle that once this PR is merged.

Red-Portal commented:
Hi @penelopeysm , what are your plans for this PR?

penelopeysm commented:
Apologies, I took an actual Christmas break and this slipped off the radar. I'll tidy it up tomorrow or something.

penelopeysm committed:
Remove Turing ext

penelopeysm commented:
> I think it is fine to remove the custom initialization.

In that case, the entire extension can be removed! Essentially, all the functions that need to be overloaded are now in AbstractMCMC, so they just live in `src`. I've updated the PR by, well, simply removing the extension.

I think in `Project.toml` the extras section for test deps is not needed too, since you have a `test/Project.toml`, so I removed those too to avoid redundant information. It all works locally for me.

Finally, I also removed the tests that used to check the behaviour of `initial_sample`.

Now unfortunately CI will still fail -- that is due to a DynamicPPL bug, which I'll fix now.

penelopeysm commented:
https://github.com/TuringLang/DynamicPPL.jl/pull/1202 should fix it, but I'll only be able to check tomorrow.

penelopeysm committed:
Enforce minimum DynamicPPL bound

Red-Portal reviewed (approved):
All looks great to me. Seems like we don't even need a DynamicPLL extension now, is that correct?

penelopeysm commented:
Yup, that's right :)

Red-Portal commented:
@penelopeysm Could you add a one-liner docstring to `AbstractStateWithTransition` so that `Documenter` doesn't complain? I'll merge the PR immediately after that.

penelopeysm commented:
Ah apologies I should have checked CI haha

penelopeysm committed:
Fix docs build

Red-Portal commented:
This is really nice. Thank you very much!
